### ISSUE TITLE
Wrong type inferrance on "and" operator

### ISSUE THREAD

**Issue by le-chartreux** [python/mypy#15809]
**Bug Report**

Hello everyone,
I encounter an 'Incompatible types in assignment' error when I run mypy on a very simple part of code concerning an assignement to a bool variable with a "and", and I think that there is no reason to raise it (I asked chatgpt and it agrees).
Thank you for your awesome work!

**To Reproduce**

Gist URL: https://gist.github.com/mypy-play/ac79e3fce931905ed5192980965e3efa
Playground URL: https://mypy-play.net/?mypy=latest&python=3.11&gist=ac79e3fce931905ed5192980965e3efa

```python
a: bool = "hello" and True
print(type(a))  # <class 'bool'>
# with mypy: main.py:1: error: Incompatible types in assignment (expression has type "str | bool", variable has type "bool")  [assignment]
```

**Expected Behavior**

I expect mypy to don't raise any error, since with a `and` operator Python should return a boolean value (correct me if I am wrong). If it had been an "or" operator I could have understood, but here I don't.

**Actual Behavior**

Taken from the mypy playground:

```
main.py:1: error: Incompatible types in assignment (expression has type "str | bool", variable has type "bool")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: latest (1.4.1) on the mypy playground
- Mypy command-line flags: I don't know, it's the default of the playground
- Mypy configuration options from `mypy.ini` (and other config files): I don't know, it's the default of the playground
- Python version used: 3.11 (playground)

Same behaviour on my Windows 10 with Python 3.11 and mypy 1.4.1.

**Comment by SanjayManjunath** [comment]
Hey @le-chartreux , try this:  https://mypy-play.net/?mypy=latest&python=3.11&gist=014dcf3563c28fe1c562bb096fad0d75

**Comment by le-chartreux** [comment]
Hello Sanjay,
Indeed, it work, thank you! But it's sad to not be able to use truthiness.

**Comment by hauntsaninja** [comment]
If the string is empty, your result will not be a bool.

**Comment by le-chartreux** [comment]
Hello Shantanu,
Indeed! I didn't know about this behaviour of Python, I was thinking that it will always be a bool. 
Sorry for this wrong bug report, and best regards.

**Comment by ikonst** [comment]
@hauntsaninja What surprises me is that even though `LiteralType.can_be_true_default` and `can_be_false_default` is defined in a way you'd expect, since we're checking `Instance.can_be_true` and it doesn't forward to the last known value, we end up not knowing what's obvious here.

**Comment by ikonst** [comment]
Or maybe the culprit here is that an 'if' statement is not considered to be "literal context"? I have a feeling all of this must've been visited and revisited already, but correct me if I'm wrong.
